### PR TITLE
Port CharacterBufferReference

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?

This brings Avalonia even more in line with the WPF implementation and makes it possible to do text layout with multiple sources of text.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026304-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026817-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026927-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027298-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+        public ReadOnlyMemory<char> CharacterBuffer { get; }
+
+        public int OffsetToFirstChar { get; }
+
+        /// <summary>
+        /// A MemoryManager over a raw pointer
+        /// </summary>
+        /// <remarks>The pointer is assumed to be fully unmanaged, or externally pinned - no attempt will be made to pin this data</remarks>
+        public sealed unsafe class UnmanagedMemoryManager<T> : MemoryManager<T>
```
Comment (maxkatz6) [inline src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:122]: Is it expected to be a public API?

File: src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            /// Create a new UnmanagedMemoryManager instance at the given pointer and size
+            /// </summary>
+            /// <remarks>It is assumed that the span provided is already unmanaged or externally pinned</remarks>
+            public UnmanagedMemoryManager(Span<T> span)
+            {
+                fixed (T* ptr = &MemoryMarshal.GetReference(span))
+                {
+                    _pointer = ptr;
+                    _length = span.Length;
+                }
```
Comment (maxkatz6) [inline src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:138]: I am not sure, is it safe? It gets fixed/pinned pointer to the span, but after "fixed" statement nothing will prevent span to be removed from stack and underlying buffer to be moved by GC.
Also, I am not sure if it's used anywhere, unlikely UnmanagedMemoryManager(T* pointer, length). So can it be just removed for safity?

Comment (maxkatz6) [reply src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:138]: With `UnmanagedMemoryManager(T* pointer, length)` caller is responsible to keep underlying buffer from being moved, so it's not a problem in this class.

File: src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:122]: No 👍

File: src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:138]: I think we can remove this wrapped and let just allow to create a buffer from ReadOnlyMemory

Comment (Gillibald) [reply src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs:138]: So we don't have to maintain such wrapper

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027328-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027329-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**